File: python/longest_happy_string.py
import heapq


class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        """
        NOTE:
        Time complexity: O(n), space complexity: O(n), where n is the sum of a, b, and c.
        Since the size of the heap is at most 3, the time complexity is O(1) for each operation.
        So, the time complexity is O(n) in total, where n is the sum of a, b, and c.
        """
        max_heap: list[tuple[int, str]] = []
        for count, char in [(a, "a"), (b, "b"), (c, "c")]:
            if count != 0:
                heapq.heappush(max_heap, (-count, char))
        result: str = ""
        while len(max_heap) > 0:
            cnt, c = heapq.heappop(max_heap)
            if len(result) > 1 and c == result[-1] and c == result[-2]:
                if len(max_heap) == 0:
                    break
                cnt_2, c_2 = heapq.heappop(max_heap)
                result += c_2
                cnt_2 += 1
                if cnt_2 < 0:
                    heapq.heappush(max_heap, (cnt_2, c_2))
            else:
                result += c
                cnt += 1
            if cnt < 0:
                heapq.heappush(max_heap, (cnt, c))
        return result


def test_longestDiverseString_case_1():
    # arrange
    a: int = 1
    b: int = 1
    c: int = 7
    expected: str = "ccaccbcc"

    # act
    solution = Solution()
    actual = solution.longestDiverseString(a, b, c)

    # assert
    assert expected == actual


def test_longestDiverseString_case_2():
    # arrange
    a: int = 7
    b: int = 1
    c: int = 0
    expected: str = "aabaa"

    # act
    solution = Solution()
    actual = solution.longestDiverseString(a, b, c)

    # assert
    assert expected == actual

File: rust/src/code/longest_happy_string.rs
struct Solution {}

impl Solution {
    // NOTE:
    // Time complexity: O(n), space complexity: O(n), where n is the sum of a, b, and c.
    // Since the size of the heap is at most 3, the time complexity is O(1) for each operation.
    // So, the time complexity is O(n) in total, where n is the sum of a, b, and c.
    pub fn longest_diverse_string(a: i32, b: i32, c: i32) -> String {
        use std::collections::BinaryHeap;

        let mut max_heap: BinaryHeap<(i32, char)> = BinaryHeap::new();
        let mut result: Vec<char> = Vec::new();
        for (count, c) in vec![(a, 'a'), (b, 'b'), (c, 'c')] {
            if count > 0 {
                max_heap.push((count, c));
            }
        }

        while let Some((mut count, c)) = max_heap.pop() {
            let result_size = result.len();
            if result_size > 1 && c == result[result_size - 1] && c == result[result_size - 2] {
                if let Some((mut count_2, c_2)) = max_heap.pop() {
                    result.push(c_2);
                    count_2 -= 1;
                    if count_2 > 0 {
                        max_heap.push((count_2, c_2));
                    }
                } else {
                    break;
                }
            } else {
                result.push(c);
                count -= 1;
            }
            if count > 0 {
                max_heap.push((count, c));
            }
        }

        result.iter().collect::<String>()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_longest_diverse_string_case_1() {
        // arrange
        let a: i32 = 1;
        let b: i32 = 1;
        let c: i32 = 7;
        let expected: String = "ccbccacc".to_string();

        // act
        let actual = Solution::longest_diverse_string(a, b, c);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_longest_diverse_string_case_2() {
        // arrange
        let a: i32 = 7;
        let b: i32 = 1;
        let c: i32 = 0;
        let expected: String = "aabaa".to_string();

        // act
        let actual = Solution::longest_diverse_string(a, b, c);

        // assert
        assert_eq!(expected, actual);
    }
}
